idaes/models_extra/power_generation/unit_models/soc_submodels.py

```diff
@@ -0,0 +1,3997 @@
+        nx = len(xnodes)
+    elif top is not None and bottom is not None:
+        nx = len(xnodes) + 2
+    else:
+        nx = len(xnodes) + 1
+
+    data_z = [None] * nx
+    data_x = [None] * nx
+    data_w = [None] * nx
+    hg = {}
+    xg = {}
+    zg = {}
+    for it, t in enumerate(time):
+        i = 0
+        zg[it] = [None] * nz
+        xg[it] = [None] * nz
+        hg[it] = [None] * nz
+        for iz, z in enumerate(znodes):
+            for ix, x in enumerate(xnodes):
+                data_z[ix] = z
+                data_x[ix] = x
+                data_w[ix] = pyo.value(var[t, ix + 1, iz + 1])
+                i += 1
+            zg[it][iz] = copy.copy(data_z)
+            xg[it][iz] = copy.copy(data_x)
+            hg[it][iz] = copy.copy(data_w)
+        # Add top and/or bottom
+        for iz, z in enumerate(znodes):
+            if bottom is not None:
+                zg[it][iz].pop()
+                xg[it][iz].pop()
+                hg[it][iz].pop()
+                zg[it][iz].insert(0, z)
+                xg[it][iz].insert(0, 0.0)
+                hg[it][iz].insert(0, pyo.value(bottom[t, iz + 1]))
+            if top is not None:
+                zg[it][iz][-1] = z
+                xg[it][iz][-1] = 1.0
+                hg[it][iz][-1] = pyo.value(top[t, iz + 1])
+        if left is not None:
+            zg[it].pop()
+            xg[it].pop()
+            hg[it].pop()
+            zg[it].insert(0, copy.copy(data_z))
+            xg[it].insert(0, copy.copy(data_z))
+            hg[it].insert(0, copy.copy(data_z))
+            if bottom is not None:
+                offset = 1
+                zg[it][0][0] = 0.0
+                xg[it][0][0] = 0.0
+                hg[it][0][0] = None
+            else:
+                offset = 0
+            if top is not None:
+                offset = 1
+                zg[it][0][-1] = 0.0
+                xg[it][0][-1] = 1.0
+                hg[it][0][-1] = None
+            for ix, x in enumerate(xnodes):
+                zg[it][0][ix + offset] = 0.0
+                xg[it][0][ix + offset] = x
+                hg[it][0][ix + offset] = pyo.value(left[t, ix + 1])
+        if right is not None:
+            zg[it][-1] = copy.copy(data_z)
+            xg[it][-1] = copy.copy(data_z)
+            hg[it][-1] = copy.copy(data_z)
+            if bottom is not None:
+                offset = 1
+                zg[it][-1][0] = 1.0
+                xg[it][-1][0] = 0.0
+                hg[it][-1][0] = None
+            else:
+                offset = 0
+            if top is not None:
+                offset = 1
+                zg[it][-1][-1] = 1.0
+                xg[it][-1][-1] = 1.0
+                hg[it][-1][-1] = None
+            for ix, x in enumerate(xnodes):
+                zg[it][-1][ix + offset] = 1.0
+                xg[it][-1][ix + offset] = x
+                hg[it][-1][ix + offset] = pyo.value(right[t, ix + 1])
+
+    return zg, xg, hg
+
+
+element_list = ["Ar","H","O","C","N","S"]
```

Comment:
This is all related to thermophysical properties, which normally would not go in a unit model file.

Reply:
Unfortunately, that's my fault.  Given the complexity of this 2D model and slowness of some general properties, I decided for efficiency and robustness to hard code the properties in this one.  It's just abnormally complicated and I wanted tight control of everything.